zaza/openstack/charm_tests/ceph/rbd_mirror/tests.py

```diff
@@ -258,85 +347,170 @@ def create_volume_from_image(cinder, image, retry=20):
 class CephRBDMirrorControlledFailoverTest(CephRBDMirrorBase):
     """Encapsulate ``ceph-rbd-mirror`` controlled failover tests."""
 
+    def cinder_fail_over_fall_back(self):
+        """Validate controlled fail over and fall back via the Cinder API."""
+        session = openstack.get_overcloud_keystone_session()
+        cinder = openstack.get_cinder_session_client(session)
+        volume = self.create_cinder_volume(session, from_image=True)
+        self.wait_for_mirror_state(
+            'up+replaying',
+            check_entries_behind_master=True,
+            application_name=self.application_name + self.site_b_app_suffix,
+            model_name=self.site_b_model,
+            pools=['cinder-ceph'])
+        self.failover_cinder_volume_host(
+            cinder_client=cinder)
+        self.assertEqual(cinder.volumes.get(volume.id).status, 'available')
+        self.failover_cinder_volume_host(
+            cinder_client=cinder,
+            target_backend_id='default',
+            target_status='enabled',
+            target_replication_status='enabled')
+        self.assertEqual(cinder.volumes.get(volume.id).status, 'available')
+
     def test_fail_over_fall_back(self):
         """Validate controlled fail over and fall back."""
```

Comment:
This is hard to understand now why this function is doing way more than just calling `cinder_fail_over_fall_back()`, although both functions have pretty much the same description. Is there a way to make things crystal clear in the description of the functions? Thanks!

EDIT: Having now read it fully, it feels like we demote, promote, resync stuff, and then finally we do a failover and then fall back. Can it be that this is a giant test validating many things? Would it be wise to split it? I'm sorry I'm having a really hard time to follow that scenario.

Reply:
I believe you're right. The function was already hard to understand as it has a lot of business logic. Adding more on top on it is not a good idea.

I'll refactor the function a bit.